### PR TITLE
Add tests and some groundwork for future contributions

### LINKED ISSUE #8: Tests Missing
The project currently has no tests. It would be good to have tests early on, as this would make for a safer and more robust development going forward.

### LINKED ISSUE #9: Magic number in SimpleCache and FileSystemCache
The [pruning](https://github.com/pallets/cachelib/blob/159ee3974bd71b939840ed302cbdf03fd7b16767/cachelib/simple.py#L36) contains a magic number, it seems it deletes only every third entry: `idx % 3`.

Does anyone know the intention behind this?

I traced it back to the very beginning, there is no info there: https://github.com/mitsuhiko/zine/commit/3502607828b0f6ed136ed9ae74a09fa10f78a3ed#diff-b0794f264f02b7241cc1088d07e16659

### LINKED ISSUE #20: Missing file count updates in `get()` and `has()`
After `os.remove()` in `FileSystemCache.get()` and `FileSystemCache.has()`, `self._update_count()` should be called.

### PR DESCRIPTION
**Pull request by northernSage**
**Edit:** I have decided to break this up into multiple PRs to make things easier for maintainers. I will close this as soon as I'm done.

Following changes were all migrated to a single "organizational" PR #24 
- [x] setup pre-commit
- [x] contribution guidelines
- [x] tox
- [x] docs (sphinx)
- [x] coverage

Following changes were moved to a single PR covering the test suite #25 
- [x] common base tests
- [x] `SimpleCache` tests
- [x] `RedisCache` tests
- [x] `MemcachedCache` Tests
- [x] `FileSystemCache` Tests

Fixes:
- [x] `cachelib.SimpleCache._prune` removing non-expired entries (migrate to PR #26)
- [x] `cachelib.FileSystemCache._prune` removing non-expired entries (migrated to PR #26) 
- [x] import `cPickle` transparently through `import pickle` avoiding `ModuleNotFoundError` for python 3 (migrated to PR #30)
- [x] `cachelib.MemcachedCache.inc` and `cachelib.MemcachedCache.dec` not initializing entries to delta or -delta as specified in `cachelib.BaseCahe` interface (migrated to PR #31)
- [x] update file count after has/get removal in `FileSystemCache` #20 (migrated to PR #33)

Fixes #8
Fixes #9 
Fixes #20

### CODE REVIEW

**Comment by northernSage** [comment]
on a side note, do we still need to support python 2? I would be glad to write a PR to drop it.

**Comment by northernSage** [comment]
I'm making this a Draft again since I have decided to start working on tests here instead of a separate PR (suggestions are welcome on this).

**Comment by northernSage** [comment]
Since there is a lot of repeated tests that need to be run for every cache type I ended up moving this code to a separate `CommonTests` base class which will be inherited by the other _cache-specific_ test classes. I'm open to suggestions if anyone has a better idea on how to go about that.

**Comment by northernSage** [comment]
```
#  cachelib.SimpleCache 

if (expires != 0 and expires <= now) or idx % 3 == 0:
    toremove.append(key)
```
This was pruning entries not yet expired (with `idx % 3 == 0`) once threshold was reached. I have changed it so it first removes all expired entries (`_remove_expired`) and if after that we are still above threshold,  older entries will be successively removed until total gets under the given threshold (`_remove_older`)

**Comment by northernSage** [comment]
Couple of things:

- from [pickle docs](https://docs.python.org/3/library/pickle.html) we can see that protocol version 5 was added in Python 3.8 so I used [`DEFAULT_PROTOCOL`](https://docs.python.org/3/library/pickle.html#pickle.DEFAULT_PROTOCOL) instead of [`HIGHEST PROTOCOL`](https://docs.python.org/3/library/pickle.html#pickle.HIGHEST_PROTOCOL) for compatibility with older python 3 releases

- `cachelib.FileSystemCache._prune` was showing the same problem as `cachelib.SimpleCache._prune` where entries not yet expired were being removed (unnecessarily as far as I can tell) when over threshold. I have changed it so it behaves like `cachelib.SimpleCache._prune` where expired entries will be removed first and after that, only if still over threshold, older entries will be sequentially removed until total is under threshold.

- due to [picke docs performance section](https://docs.python.org/3/library/pickle.html#performance) and [pickle source](https://github.com/python/cpython/blob/5b22dd87aa39087f07987f788a0bbd2464e2a8b5/Lib/pickle.py#L1348-L1352) importing cPickle directly isn't necessary for python 3 and will raise `ModuleNotFoundError`, so the code now imports pickle normally and cPickle will be transparently imported.

**Comment by northernSage** [comment]
`cachelib.MemcachedCache.inc` and `cachelib.MemcachedCache.dec` were not initializing inexistent entries to `delta` or `-delta` as specified in the `cachelib.BaseCahe` interface and consequently failing tests, so I made some small adjustments.

**Comment by northernSage** [comment]
`Memcached` tests as showing inconsistent results. I will work on it some more and open a new PR in the future. For now, this is ready for review.

**Comment by lowne** [comment]
> This was pruning entries not yet expired (with `idx % 3 == 0`) once threshold was reached. I have changed it so it first removes all expired entries (`_remove_expired`) and if after that we are still above threshold, older entries will be successively removed until total gets under the given threshold (`_remove_older`)

@northernSage it *might* be better performance wise (genuinely not sure) to only iterate over the entries once - something like https://github.com/sh4nks/flask-caching/issues/155#issuecomment-717239768

That approach also respects using different timeouts within a cache (i.e. longer timeouts for more expensive entries), pruning the cheap entries first; and adds a configuration option to tune the frequency of prune events.

File: src/cachelib/file.py [hunk 1]
```diff
@@ -75,60 +70,85 @@ def _normalize_timeout(self, timeout):
...
+        for fname in entries:
+            try:
+                with open(fname, "rb") as f:
+                    exp_fname_tuples.append((pickle.load(f), fname))
+            except OSError:
+                pass
+        fname_sorted = [
+            fname for _, fname in sorted(exp_fname_tuples, key=lambda item: item[1][0])
+        ]
+        for fname in fname_sorted:
```
Comment (lowne) [inline src/cachelib/file.py:110]: isn't this going to nuke the whole thing?

Comment (northernSage) [reply src/cachelib/file.py:110]: Yep, nice catch. Fixed.

**Comment by northernSage** [comment]
Hey, it's nice to see someone else interested in this. It has been a while since I wrote the PR so I had to go through the code again before getting back to you. I was working on something else so it took me a few days to get around to it, sorry about the delay.

> @northernSage it _might_ be better performance wise (genuinely not sure) to only iterate over the entries once - something like [sh4nks/flask-caching#155 (comment)](https://github.com/sh4nks/flask-caching/issues/155#issuecomment-717239768)
> 
> That approach also respects using different timeouts within a cache (i.e. longer timeouts for more expensive entries), pruning the cheap entries first; and adds a configuration option to tune the frequency of prune events.

Yes, this can definitely be optimized. That being said, when I wrote this, performance was not my number one concern. The idea of the implementation is providing a consistent, simple and predictable behavior. The main purpose of this PR is to prepare the project for future contributions (tests, pre-commit hooks, etc). I only changed the prune functionality because it was making the test suite flaky. Optimization can come at a later stage with separate PRs.

On a side note, if someone is worried about performance, a filesystem-based cache which operates on disk would be a less than ideal choice. The person would probably be better served with an in-memory cache system, such as Redis.

Thanks for pointing out the bug in `_remove_older`, though. I totally missed that. In case you have any other suggestions, please let me know. 👍🏻

**Comment by odidev** [comment]
@northernSage  @lowne 
I am working with cachelib to build and test it on the linux platform.
I can see that a [PR](https://github.com/pallets/cachelib/pull/24 ) has been merged yesterday, which describes how to build and test cachelib. However, no test was running during the pytest.
I have pulled this PR #13 (To add tests in cachelib repo) locally and tested it. Out of 44 tests, 24 have passed but 20 are getting skipped. Below is the test summary:

~~~
========================================================== test session starts ==========================================================
platform linux -- Python 3.8.3, pytest-6.2.4, py-1.9.0, pluggy-0.13.1
rootdir: /workspace/cachelib, configfile: setup.cfg, testpaths: tests
plugins: xprocess-0.17.1
collected 44 items

tests/test_file_system_cache.py .............                                                                                     [ 29%]
tests/test_memcached_cache.py ssssssssss                                                                                          [ 52%]
tests/test_redis_cache.py ssssssssss                                                                                              [ 75%]
tests/test_simple_cache.py ...........                                                                                            [100%]

==================================================== 24 passed, 20 skipped in 5.62s =====================================================
~~~

The logs do not show any possible reason for the 20 tests that skipped. However, it seems that the test suite requires redis-server to be installed and in active state. I have installed the redis-server and activated the redis-server service, but these test cases are still getting skipped.

Can you please provide me pointers on how to run those skipped test cases, and their requirements?
Also, may I know about the plan to push this PR further to add tests in this repo?

**Comment by northernSage** [comment]
Hey @odidev, thanks for taking the time to point out the issue. Cachelib has been under active reorganization/development in the last couple days and I should have made that more clear by mentioning it somewhere more visible, sorry about that. This PR is currently being broken down into smaller chunks which will be revised and merged once maintainers find it acceptable. The tests have been moved to #25 and should be merged soon once a few issues regarding #26 have been sorted out. Finally, the tests are most likely being skipped due to missing dependencies: [redis-py](https://github.com/andymccurdy/redis-py) and [pylibmc](http://sendapatch.se/projects/pylibmc/).

**Comment by northernSage** [comment]
Closed in favour of #24, #25, #26, #30, #31, #32, #33

**Pull request closed without merging**